src/server/controlchan/commands/pass.rs

```diff
@@ -73,23 +73,56 @@ where
                     source_ip: session.source.ip(),
                     certificate_chain: session.cert_chain.clone(),
                 };
+                let failedlogins = session.failedlogins.clone();
+                let source_ip = session.source.ip();
                 tokio::spawn(async move {
```

Comment:
This method is getting rather long, it will probably benefit from splitting some things into functions.

Reply:
Yes I totally agree. This is getting too complicated and therefore risky.
I've also now inserted some failed login attempt delays, that we should actually move away here. They are also implemented in the jsonfile auth crate.
I was thinking to add some more unit and integration tests to make sure it all still does what it's supposed to. But we need to refactor this.